File: android/app/src/main/kotlin/net/chikach/submon/mch/MessagingMethodChannelHandler.kt
package net.chikach.submon.mch

import android.Manifest
import android.os.Build
import com.google.firebase.messaging.FirebaseMessaging
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import net.chikach.submon.MainActivity

const val REQUEST_CODE_NOTIFICATION_PERMISSION = 16

class MessagingMethodChannelHandler(private val activity: MainActivity) : MethodChannel.MethodCallHandler {
    var methodResult: MethodChannel.Result? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "requestNotificationPermission" -> {
                requestNotificationPermission(result)
            }
            "getToken" -> {
                getToken(result)
            }
            else -> result.notImplemented()
        }
    }

    private fun requestNotificationPermission(result: MethodChannel.Result) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            activity.requestPermissions(arrayOf(Manifest.permission.POST_NOTIFICATIONS), REQUEST_CODE_NOTIFICATION_PERMISSION)
            methodResult = result
        } else {
            result.success(NotificationPermissionState.GRANTED.ordinal)
        }
    }

    fun completeRequestNotificationPermission(granted: Boolean) {
        if (granted) {
            methodResult?.success(NotificationPermissionState.GRANTED.ordinal)
        } else {
            methodResult?.success(NotificationPermissionState.DENIED.ordinal)
        }
    }

    private fun getToken(result: MethodChannel.Result) {
        FirebaseMessaging.getInstance().token.addOnSuccessListener {
            result.success(it)
        }.addOnFailureListener {
            result.error("unknown", it.message, null)
        }
    }

    enum class NotificationPermissionState {
        GRANTED,
        DENIED,
    }
}

File: ios/Runner/ChannelHandler/MessagingMethodChannelHandler.swift
import Foundation
import FirebaseMessaging

class MessagingMethodChannelHandler {
    let methodChannel: FlutterMethodChannel
    let viewController: FlutterViewController
    let appDelegate: AppDelegate
    
    init(viewController: FlutterViewController, appDelegate: AppDelegate) {
        self.methodChannel = FlutterMethodChannel(name: "net.chikach.submon/messaging", binaryMessenger: viewController.binaryMessenger)
        self.viewController = viewController
        self.appDelegate = appDelegate
    }
    
    func register() {
        methodChannel.setMethodCallHandler(handler)
    }
    
    private func handler(call: FlutterMethodCall, result: @escaping FlutterResult) {
//        let args = call.arguments as! [String: String]
        
        switch call.method {
        case "requestNotificationPermission":
            requestNotificationPermission(result: result)
            break
        case "getToken":
            Messaging.messaging().token(completion: { (token, error) in
                if error != nil {
                    result(nil)
                } else if let token = token {
                    result(token)
                }
            })
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func requestNotificationPermission(result: @escaping FlutterResult) {
        let notificationCenter = UNUserNotificationCenter.current()
        
        notificationCenter.requestAuthorization(options: [.alert, .sound, .badge]) { (granted, _) in
            if granted {
                notificationCenter.delegate = self.appDelegate
                result(NotificationPermissionState.granted.rawValue)
            } else {
                result(NotificationPermissionState.denied.rawValue)
            }
        }
    }
}

enum NotificationPermissionState: Int {
    case granted = 0
    case denied = 1
}
